pkg/observable/channel/replay.go

```diff
@@ -0,0 +1,209 @@
+	var values []V
+	// accumulateReplayValues works concurrently and returns a context and cancellation
+	// function for signaling completion.
+	valuesAccCtx, valuesAccCancel := accumulateReplayValues(tempObserver, n, &values)
+
+	// Wait for N values to be accumulated or timeout. When timing out, if we
+	// have at least 1 value, we can return it. Otherwise, we need to wait for
+	// the next value to be published (i.e. continue the loop).
+	for {
+		select {
+		case <-valuesAccCtx.Done():
+			return values
+		case <-time.After(replayPartialBufferTimeout):
+			if len(values) > 1 {
+				valuesAccCancel()
+				return values
+			}
+		}
+	}
+}
+
+// Subscribe returns an observer which is notified when the publishCh channel
+// receives a value.
+func (ro *replayObservable[V]) Subscribe(ctx context.Context) observable.Observer[V] {
+	ro.replayBufferMu.RLock()
+	defer ro.replayBufferMu.RUnlock()
+
+	observer := NewObserver[V](ctx, ro.onUnsubscribe)
+
+	// ToReplayObservable all buffered replayBuffer to the observer channel buffer before
+	// any new values have an opportunity to send on observerCh (i.e. appending
+	// observer to ro.observers).
+	//
+	// TODO_IMPROVE: this assumes that the observer channel buffer is large enough
+	// to hold all replay (buffered) notifications.
+	for _, notification := range ro.replayBuffer {
+		observer.notify(notification)
+	}
+
+	ro.addObserver(observer)
+
+	return observer
+}
+
+// UnsubscribeAll unsubscribes and removes all observers from the observable.
+func (ro *replayObservable[V]) UnsubscribeAll() {
+	ro.unsubscribeAll()
+}
+
+// goBufferReplayNotifications buffers the last n notifications from a source
+// observer. It is intended to be run in a goroutine.
+func (ro *replayObservable[V]) goBufferReplayNotifications(srcObserver observable.Observer[V]) {
+	for notification := range srcObserver.Ch() {
+		ro.replayBufferMu.Lock()
+		// Add the notification to the buffer.
+		if len(ro.replayBuffer) < ro.replayBufferSize {
+			ro.replayBuffer = append(ro.replayBuffer, notification)
+		} else {
+			// buffer full, make room for the new notification by removing the
+			// oldest notification.
+			ro.replayBuffer = append(ro.replayBuffer[1:], notification)
+		}
+		ro.replayBufferMu.Unlock()
+	}
+}
+
+// accumulateReplayValues asynchronously accumulates n values from the observer channel
+// into the slice pointed to by accValues.
+// It returns a context and corresponding cancellation function. It cancels the
+// context when n values have been accumulated. Callers SHOULD also call the
+// cancellation function, if the context isn't done, before using the accumulated
+// values.
+func accumulateReplayValues[V any](
+	observer observable.Observer[V],
+	n int, accValues *[]V,
+) (context.Context, context.CancelFunc) {
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		for {
+			// The context was cancelled since the last iteration.
```

Comment:
```suggestion
			// The context was canceled since the last iteration.
```

Reply:
I suppose my bias is exposed: https://www.grammarly.com/blog/canceled-vs-cancelled/

Reply:
Well, this is akward:

![image](https://github.com/pokt-network/poktroll/assets/600733/8bc891e0-96eb-4de6-8b2a-1b5c443df556)